esutil/bulk_indexer.go

```diff
@@ -0,0 +1,509 @@
+	numCreated  uint64
+	numUpdated  uint64
+	numDeleted  uint64
+	numRequests uint64
+}
+
+// NewBulkIndexer creates a new bulk indexer.
+//
+func NewBulkIndexer(cfg BulkIndexerConfig) (BulkIndexer, error) {
+	if cfg.Client == nil {
+		cfg.Client, _ = elasticsearch.NewDefaultClient()
+	}
+
+	if cfg.Decoder == nil {
+		cfg.Decoder = defaultJSONDecoder{}
+	}
+
+	if cfg.NumWorkers == 0 {
+		cfg.NumWorkers = runtime.NumCPU()
+	}
+
+	if cfg.FlushBytes == 0 {
+		cfg.FlushBytes = 5e+6
+	}
+
+	if cfg.FlushInterval == 0 {
+		cfg.FlushInterval = 10 * time.Second
+	}
+
+	bi := bulkIndexer{
+		config: cfg,
+		stats:  &bulkIndexerStats{},
+	}
+
+	bi.init()
+
+	return &bi, nil
+}
+
+// Add adds an item to the indexer.
+//
+func (bi *bulkIndexer) Add(ctx context.Context, item BulkIndexerItem) error {
+	atomic.AddUint64(&bi.stats.numAdded, 1)
+
+	select {
+	case <-ctx.Done():
+		return ctx.Err()
+	case bi.queue <- item:
+	}
+
+	return nil
+}
+
+// Close calls flush on writers and closes the indexer queue channel.
+//
+func (bi *bulkIndexer) Close(ctx context.Context) error {
+	select {
+	case <-ctx.Done():
+		return ctx.Err()
+	default:
+		close(bi.queue)
+		bi.wg.Wait()
+	}
+
+	bi.ticker.Stop()
+
+	for _, w := range bi.workers {
+		w.mu.Lock()
+		if w.buf.Len() > 0 {
+			if err := w.flush(); err != nil {
+				// TODO(karmi): Wrap error
+				w.mu.Unlock()
+				return fmt.Errorf("%s", err)
+			}
+			w.mu.Unlock()
+		}
+	}
+
+	return nil
+}
+
+// Stats returns indexer statistics.
+//
+func (bi *bulkIndexer) Stats() BulkIndexerStats {
+	return BulkIndexerStats{
+		NumAdded:    uint(atomic.LoadUint64(&bi.stats.numAdded)),
+		NumFlushed:  uint(atomic.LoadUint64(&bi.stats.numFlushed)),
+		NumFailed:   uint(atomic.LoadUint64(&bi.stats.numFailed)),
+		NumIndexed:  uint(atomic.LoadUint64(&bi.stats.numIndexed)),
+		NumCreated:  uint(atomic.LoadUint64(&bi.stats.numCreated)),
+		NumUpdated:  uint(atomic.LoadUint64(&bi.stats.numUpdated)),
+		NumDeleted:  uint(atomic.LoadUint64(&bi.stats.numDeleted)),
+		NumRequests: uint(atomic.LoadUint64(&bi.stats.numRequests)),
+	}
+}
+
+// init initializes the bulk indexer.
+//
+func (bi *bulkIndexer) init() {
+	bi.queue = make(chan BulkIndexerItem, bi.config.NumWorkers)
+
+	for i := 1; i <= bi.config.NumWorkers; i++ {
+		w := worker{
+			id:  i,
+			ch:  bi.queue,
+			bi:  bi,
+			buf: bytes.NewBuffer(make([]byte, 0, bi.config.FlushBytes)),
+			aux: make([]byte, 0, 512)}
+		w.run()
+		bi.workers = append(bi.workers, &w)
+	}
+	bi.wg.Add(bi.config.NumWorkers)
+
+	bi.ticker = time.NewTicker(bi.config.FlushInterval)
+	go func() error {
+		for {
+			select {
+			case <-bi.ticker.C:
+				if bi.config.DebugLogger != nil {
+					bi.config.DebugLogger.Printf("[indexer] Auto-flushing workers after %s\n", bi.config.FlushInterval)
+				}
+				for _, w := range bi.workers {
+					w.mu.Lock()
+					if w.buf.Len() > 0 {
+						if err := w.flush(); err != nil {
+							// TODO(karmi): Wrap error
+							w.mu.Unlock()
+							return fmt.Errorf("%s", err)
```

Comment:
This is going to cause the periodic flusher to exit on the first error - is that desirable? What about transient errors, like a network outage? Probably not great behaviour for a long-lived bulk indexer.

Reply:
No, it wouldn't stop the entire indexer - but the periodic flusher would no longer operate. Seems wrong to me to leave the indexer in a half-working state.

Reply:
That is correct — thanks for raising it, it's a buggy behaviour.

Reply:
moving flush to the worker would help/solve this.

Reply:
This had been removed in 9ec0d6b: 
https://github.com/elastic/go-elasticsearch/blob/9ec0d6bcffba0fe59174febe0e7c15ae8d58521f/esutil/bulk_indexer.go#L294-L300